pkg/urpc/urpc.go

```diff
@@ -89,6 +89,35 @@ func (f *FilePayload) setFilePayload(fs []*os.File) {
 	f.Files = fs
 }
 
+// ReadonlyFilePayload is the same as FilePayload, except that it does not
+// export the underlying files array. Using an unexported type alias, it can be
+// embedded without being assignable directly outside a package. This is useful
+// for maintaining invariants over the files and other struct fields.
+type ReadonlyFilePayload struct {
+	files []*os.File
+}
```

Comment:
This is very cool, but does the use-case justify the added complexity?

It seems rare that we reach into `Files[]` after the initial construction.  `Exec` is probably the only one that does that.  We could replace this with a comment on `Exec.GuestFDs` that explains the 1:1 mapping with `Files[]`, and return an erorr in the `Exec` handler if they are not equal length.

I don't feel too strongly about this, so if you think it's worth keeping the `ReadonlyFilePayload` that is fine with me.

Reply:
As there are other places where it is easy to violate similar constraints (cf. `boot.StartArgs` and `NumOverlayFilestoreFDs` that partitions the transmitted files), it is probably a bit overkill. I have removed it now.